analyzer/runtime/extract.go

```diff
@@ -364,15 +364,40 @@ func ExtractRound(blockHeader nodeapi.RuntimeBlockHeader, txrs []nodeapi.Runtime
 					blockTransactionData.Body = body
 					amount = body.Amount.Amount
 					if body.To != nil {
-						// Beware, this is the address of an account in the consensus
-						// layer, not an account in the runtime that generated this event.
+						// This is the address of an account in the consensus
+						// layer, not an account in the runtime that generated this tx.
 						// We do not register it as a preimage.
 						if to, err = uncategorized.StringifySdkAddress(body.To); err != nil {
 							return fmt.Errorf("to: %w", err)
 						}
 					}
 					return nil
 				},
+				ConsensusAccountsDelegate: func(body *consensusaccounts.Delegate) error {
+					blockTransactionData.Body = body
+					amount = body.Amount.Amount
+					// This is the address of an account in the consensus
+					// layer, not an account in the runtime that generated this tx.
+					// We do not register it as a preimage.
+					if to, err = uncategorized.StringifySdkAddress(&body.To); err != nil {
+						return fmt.Errorf("to: %w", err)
+					}
+					return nil
+				},
+				ConsensusAccountsUndelegate: func(body *consensusaccounts.Undelegate) error {
+					blockTransactionData.Body = body
+					// NOTE: The `from` and `to` addresses have swapped semantics compared to most other txs:
+					// Assume R is a runtime address and C is a consensus address. The inverse of Delegate(from=R, to=C) is Undelegate(from=C, to=R).
+					if _, err := registerRelatedSdkAddress(blockTransactionData.RelatedAccountAddresses, &body.From); err != nil {
+						return fmt.Errorf("from: %w", err)
+					}
+					// The `to` is implicitly the account that created this tx. Ref: https://github.com/oasisprotocol/oasis-sdk/blob/eb97a8162f84ae81d11d805e6dceeeb016841c27/runtime-sdk/src/modules/consensus_accounts/mod.rs#L465-L465
+					to = blockTransactionData.SignerData[0].Address
+					// The `amount` (of tokens) is not contained in the body, only `shares` is. We do not have sufficient information
+					// to convert `shares` to `amount`; we'd need the total amount of shares and tokens in the `from` account.
+					// Assign nothing to `amount`.
```

Comment:
Can anybody confirm this?

Also, if true, should we store the number of shares in the `amount` variable even though the semantics are different? Both `to` and `amount` are loosely defined and introduced by the indexer; they're a kind of best-effort intentionally-over-generalized fields to help navigate the txs quickly almost regardless of their call method.

The `amount` variable contents end up in the `amount` column in our DB, and eventually in the `amount` field of each tx in the API, but the API docs describe the above looseness pretty well:
> A reasonable "amount" associated with this transaction, if applicable. The meaning varies based on the transaction method.

Still, might be too much of a stretch? And potentially particularly confusing because shares and token amounts are so closely related?

Reply:
Do not conflate shares and amount as this can lead to incorrect values and user confusion.

You can only know the final amount once the undelegation completes as nodes may be slashed while debonding. However you can compute the amount at any point in time by inspecting the relevant share pool (see the CLI).

Reply:
Thanks. I kept the shares in the `amount` field, as was coded initially; I agree we better be careful here.
I updated the comment with your explanation about slashing.